pysph/sph/swe/basic.py

```diff
@@ -0,0 +1,1677 @@
+import numpy as np
+
+
+class CheckForParticlesToSplit(Equation):
+    r"""Particles are tagged for splitting if the following condition is
+    satisfied:
+
+    .. math::
+        (A_i > A_max) and (h_i < h_max) and (x_min < x_i < x_max) and (y_min <
+        y_i < y_max)
+
+    References
+    ----------
+    .. [Vacondio2010] R. Vacondio, B.D. Rodgers and P.K. Stansby, "Accurate
+    particle splitting for smoothed particle hydrodynamics in shallow water
+    with shock capturing", Int. J. Numer. Meth. Fluids, 69 (2012), pp.
+    1377-1410
+
+    """
+    def __init__(self, dest, h_max=1e9, A_max=1e9, x_min=-1e9, x_max=1e9,
+                 y_min=-1e9, y_max=1e9):
+        r"""
+        Parameters
+        ----------
+        h_max : float
+            maximum smoothing length beyond which splitting is deactivated
+        A_max : float
+            maximum area beyond which splitting is activated
+        x_min : float
+            minimum distance along x-direction beyond which splitting is
+            activated
+        x_max : float
+            maximum distance along x-direction beyond which splitting is
+            deactivated
+        y_min : float
+            minimum distance along y-direction beyond which splitting is
+            activated
+        y_max : float
+            maximum distance along y-direction beyond which splitting is
+            deactivated
+
+        """
+        self.A_max = A_max
+        self.h_max = h_max
+        self.x_min = x_min
+        self.x_max = x_max
+        self.y_min = y_min
+        self.y_max = y_max
+        super(CheckForParticlesToSplit, self).__init__(dest, None)
+
+    def initialize(self, d_pa_to_split, d_idx):
+        d_pa_to_split[d_idx] = 0.0
+
+    def post_loop(self, d_idx, d_rho0, d_A, d_h, d_x, d_y, d_pa_to_split):
+        if (d_A[d_idx] > self.A_max and d_h[d_idx] < self.h_max
+           and (self.x_min < d_x[d_idx] < self.x_max)
+           and (self.y_min < d_y[d_idx] < self.y_max)):
+            d_pa_to_split[d_idx] = 1
+
+
+class ParticleSplit(object):
+    r"""**Hexagonal particle splitting algorithm**
+
+    References
+    ----------
+    .. [Vacondio2010] R. Vacondio, B.D. Rodgers and P.K. Stansby, "Accurate
+    particle splitting for smoothed particle hydrodynamics in shallow water
+    with shock capturing", Int. J. Numer. Meth. Fluids, 69 (2012), pp.
+    1377-1410
+
+    """
+    def __init__(self, pa_arr):
+        r"""
+        Parameters
+        ----------
+        pa_arr : pysph.base.particle_array.ParticleArray
+            particle array of fluid
+
+        """
+        self.pa_arr = pa_arr
+
+        # Ratio of mass of daughter particle located at center of hexagon to
+        # that of its parents mass
+        self.center_pa_mass_frac = 0.178705766141917
+
+        # Ratio of mass of daughter particle located at edge of hexagon to that
+        # of its parents mass
+        self.edge_pa_mass_frac = 0.136882287617319
```

Comment:
Why the term edge?  Why not just daughter?

Reply:
I have changed this to vertex_pa instead of edge_pa. There are actually daughter particles located at the center of the hexagon and vertices of the hexagon after splitting, so to distinguish between them I used these names.